tactical-microgrid-standard/common/TimerHandler.h

```diff
@@ -0,0 +1,246 @@
+#ifndef TMS_TIMER_HANDLER_H
+#define TMS_TIMER_HANDLER_H
+
+#include <ace/Event_Handler.h>
+#include <ace/Log_Msg.h>
+#include <ace/Reactor.h>
+
+#include <chrono>
+#include <map>
+#include <variant>
+#include <mutex>
+#include <type_traits>
+#include <memory>
+#include <stdexcept>
+
+using Sec = std::chrono::seconds;
+using Clock = std::chrono::system_clock;
+using TimePoint = std::chrono::time_point<Clock>;
+using TimerId = long;
+constexpr TimerId null_timer_id = 0;
+using Mutex = std::recursive_mutex;
+using Guard = std::lock_guard<Mutex>;
+
+template <typename EventType>
+struct Timer {
+  using Arg = EventType;
+  using Ptr = std::shared_ptr<Timer<EventType>>;
+  using Map = std::map<std::string, Ptr>;
+  using MapPtr = std::shared_ptr<Map>;
+
+  std::string name;
+  TimerId id = null_timer_id;
+  EventType arg = {};
+  Sec period = Sec(0);
+  Sec delay = Sec(0);
+  bool exit_after = false;
+
+  bool active() const
+  {
+    return id != null_timer_id;
+  }
+};
+
+template <typename EventType>
+class TimerHolder {
+public:
+  TimerHolder()
+    : timers_(std::make_shared<typename Timer<EventType>::Map>())
+  {
+  }
+
+  explicit operator typename Timer<EventType>::MapPtr()
+  {
+    return timers_;
+  }
+
+private:
+  typename Timer<EventType>::MapPtr timers_;
+};
+
+/**
+ * Wrapper around ACE_Event_Handler to make using a fixed set of timers easier
+ * to use.
+ */
+template <typename... EventTypes>
+class TimerHandler : public ACE_Event_Handler, protected TimerHolder<EventTypes>...
+{
+public:
+  using AnyTimer = std::variant<typename Timer<EventTypes>::Ptr...>;
+
+  TimerHandler(ACE_Reactor* reactor = nullptr)
+    : reactor_(reactor ? reactor : ACE_Reactor::instance())
+  {
+  }
+
+  virtual ~TimerHandler()
+  {
+    cancel_all();
+  }
+
+  template <typename EventType>
+  typename Timer<EventType>::MapPtr get_timers()
+  {
+    return static_cast<typename Timer<EventType>::MapPtr>(*this);
+  }
+
+  template <typename EventType>
+  typename Timer<EventType>::Ptr get_timer(const std::string& name = "")
+  {
+    auto& timers = *get_timers<EventType>();
+    const auto it = timers.find(name);
+    if (it == timers.end()) {
+      auto timer = std::make_shared<Timer<EventType>>();
+      timer->name = name;
+      timers[name] = timer;
+      return timer;
+    }
+    return it->second;
+  }
+
+  template <typename EventType>
+  void assert_inactive(typename Timer<EventType>::Ptr timer)
+  {
+    if (timer->active()) {
+      throw std::runtime_error("Can't schedule, timer already active!");
+    }
+  }
```

Comment:
Some functions look like they are only used internally and can be private or protected. For example, this function and `timer_wont_run`.